src/bytes_mut.rs

```diff
@@ -644,21 +644,37 @@ impl BytesMut {
 
                     self.ptr = vptr(ptr);
                     self.cap = v.capacity();
+                } else {
+                    // calculate offset
+                    let off = v.capacity() - self.cap;
+
+                    // new_cap is calculated in terms of `BytesMut`, not the underlying
+                    // `Vec`, so it does not take the offset into account.
+                    //
+                    // Thus we have to manually add it here.
+                    new_cap = new_cap.checked_add(off).expect("overflow");
+
+                    // The vector capacity is not sufficient. The reserve request is
+                    // asking for more than the initial buffer capacity. Allocate more
+                    // than requested if `new_cap` is not much bigger than the current
+                    // capacity.
+                    //
+                    // There are some situations, using `reserve_exact` that the
+                    // buffer capacity could be below `original_capacity`, so do a
+                    // check.
+                    let double = v.capacity().checked_shl(1).unwrap_or(new_cap);
+
+                    new_cap = cmp::max(cmp::max(double, new_cap), original_capacity);
```

Comment:
We should only need the inner `cmp::max` call because `original_capacity` is going to match `v.capacity()`, and we're only here if our underlying capacity is already smaller than our desired capacity.

Reply:
Fixed.